### PR TITLE
Enable virtio-fs on s390x

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
- Use updated GoVMM following the merge of https://github.com/kata-containers/govmm/pull/161 and allow `appendVhostUserDevice` on s390x.
- One thing I'm not so sure about is checking for virtio-fs support. It was only added to s390x in QEMU 5.2.0, which is officially to be used with Kata since the merge of #1349, but which generally hasn't landed in distro releases yet.
We could just assume that a recent enough QEMU is used because Kata's tests/CI use it. If we do this and try virtio-fs on a non-virtio-fs QEMU, the user would receive a message like:
```
ctr: failed to launch qemu: exit status 1, error messages from qemu log: qemu-system-s390x: -device vhost-user-fs-ccw,chardev=char-50b9c9fc5d2491f7,tag=kataShared: 'vhost-user-fs-ccw' is not a valid device model name
```
even though this could be wrapped.
Alternatively, we could check whether the installed QEMU supports it, which would require moving the QEMU installation instructions in the Developer's Guide before the Kata installation.
The more stable way to do this is QAPI, which involves running an "empty" machine and querying it. This would mean the additional `socat` dependency, which would have to be added in https://github.com/kata-containers/tests. Depending on the exact setup, it may also require root access. This is the way that I have included in this PR.
The less stable way to do this is checking the output of `qemu-system-s390x -device help`.
WDYT?

Depends-on: https://github.com/kata-containers/tests/pull/3382
because currently, on s390x, packaged QEMU (=> <5.2) is preferred.
Fixes: #1469
Backport only worth the effort if v1 moves forward to QEMU 5.2 as well (is at 5.0), so no backport required, at least for the time being.

### CODE REVIEW

File: docs/Developer-Guide.md [hunk 1]
```diff
@@ -85,6 +85,46 @@ You need to install the following to build Kata Containers components:
 - `make`.
 - `gcc` (required for building the shim and runtime).
 
+# Install a hypervisor
+
```
Comment (c3d) [inline docs/Developer-Guide.md:89]: Could you split this commit in two, one with the code move, the other with the actual text changes? As is, it's dififcult to review what you actually changed.

File: docs/Developer-Guide.md [hunk 2]
```diff
@@ -85,6 +85,46 @@ You need to install the following to build Kata Containers components:
...
+When setting up Kata using a [packaged installation method](install/README.md#installing-on-a-linux-system), the `qemu-lite` hypervisor is installed automatically. For other installation methods, you will need to manually install a suitable hypervisor.
+
+## Build a custom QEMU
```
Comment (c3d) [inline docs/Developer-Guide.md:92]: Do any of these changes actually relate to s390? I did not find anything (but see earlier comment).

File: src/runtime/Makefile [hunk 3]
```diff
@@ -439,7 +453,7 @@ USER_VARS += DEFDISABLEBLOCK
 USER_VARS += DEFBLOCKSTORAGEDRIVER_ACRN
 USER_VARS += DEFBLOCKSTORAGEDRIVER_FC
 USER_VARS += DEFBLOCKSTORAGEDRIVER_QEMU
-USER_VARS += DEFSHAREDFS_QEMU_VIRTIOFS
```
Comment (c3d) [inline src/runtime/Makefile:442]: This change may break distro builds that rely on the old name. But I agree it's an improvement. I would just put it in some separate commit, because it's not obviously related to s390.

**Review by c3d — APPROVED** [review]
Approved, with comments.

**Comment by Jakob-Naucke** [comment]
- Added forgotten revendor of GoVMM
- Split detection of s390x virtio-fs availability, rename of `DEFSHAREDFS`, and move of QEMU documentation in separate commits.

File: src/runtime/Makefile (hunk 3 above)
Comment (Jakob-Naucke) [reply src/runtime/Makefile:442]: Done!

File: docs/Developer-Guide.md (hunk 1 above)
Comment (Jakob-Naucke) [reply docs/Developer-Guide.md:89]: Done. Made the rationale per commit a little clearer.

File: src/runtime/Makefile [hunk 4]
```diff
@@ -168,7 +168,21 @@ DEFAULTEXPFEATURES := []
...
+            qmp_sock=/tmp/qemu-capabilities.monitor.sock; \
+            pidfile=/tmp/qemu-capabilities.pidfile; \
+            $(QEMUPATH) -S -no-user-config -nodefaults -nographic -M none -daemonize \
+                        -qmp unix:$$qmp_sock,server,nowait -pidfile $$pidfile; \
+            echo '{"execute": "qmp_capabilities"} {"execute": "qom-list-types"}' | socat - UNIX:$$qmp_sock | \
+                tail -1 | jq ".return | .[].name" | grep -F vhost-user-fs-ccw; \
+            kill $$(cat $$pidfile); \
+            rm -f $$qmp_sock $$pidfile))
+        DEFSHAREDFS := virtio-9p
+    endif
```
Comment (fidencio) [inline src/runtime/Makefile:184]: @Jakob-Naucke, please, bear with me here and let me see if I properly understand what you're doing here.

You define virtio-fs a the default fs, then you check, for s390x, whether the QEMU on the machine used for building kata-containers, has support to virtio-fs and, if not, you fallback to virtio-9p.

Do I get it correctly?  If so, what ensures that the machine where kata-containers will be running will have a QEMU as new as the one present in the machine where it was built?

I see this as something a distro should ensure that will work, or having an "on the fly" fallback mechanism as part of the runtime code.  But not something that can be ensure in at the build time.

WDYT?

**Comment by fidencio** [comment]
Let me just add the `do-not-merge` label for now.

File: src/runtime/Makefile (hunk 4 above)
Comment (Jakob-Naucke) [reply src/runtime/Makefile:184]: > You define virtio-fs a the default fs, then you check, for s390x, whether the QEMU on the machine used for building kata-containers, has support to virtio-fs and, if not, you fallback to virtio-9p.
> 
> Do I get it correctly?

Yes, that is what I'm doing here.

>  If so, what ensures that the machine where kata-containers will be running will have a QEMU as new as the one present in the machine where it was built?

Indeed, I had not considered the possibility of having a build machine with a QEMU setup different from the target machine.

> I see this as something a distro should ensure that will work, or having an "on the fly" fallback mechanism as part of the runtime code. But not as something that can be ensured at the build time.
> 
> WDYT?

I'd be against falling back to 9p in a way that isn't very clear to the user when virtio-fs is configured, i.e. a journal entry wouldn't be enough.
I don't know whether there is a surefire way to do that except for fatal messages, so maybe I'd keep it a fatal message.
Then I'd probably want to replace what would happen now when you'd run virtio-fs without support in QEMU (`'vhost-user-fs-ccw' is not a valid device model name`, see above) with a more understandable message about virtio-fs not being available.

Does this sound reasonable?

**Comment by c3d** [comment]
I think that @fidencio has a very good point. I still like the idea of a query, but you would really need the runtime to do it.

File: src/runtime/Makefile (hunk 4 above)
Comment (fidencio) [reply src/runtime/Makefile:184]: That sounds super reasonable, and I also have the preference for not silently falling back to 9p.
A huge fatal message, and a better error report is, indeed, the way to go.

Thanks, @Jakob-Naucke!

**Comment by Jakob-Naucke** [comment]
Removed QEMU checks -- potential Kata/QEMU incompatibilities are a separate and much more general issue (see #1709).
@fidencio release gate would be a nice to have.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @Jakob-Naucke!

**Comment by fidencio** [comment]
ARM failure:
```
Comparing merged pages between containers
ERROR: The merged pages on the second container is less than the first container
time="2021-04-20T17:45:30Z" level=warning msg="task test exit with non-zero exit code 9"
time="2021-04-20T17:45:30Z" level=warning msg="task test1 exit with non-zero exit code 9"
restoring KSM settings
Makefile:127: recipe for target 'ksm' failed
make: *** [ksm] Error 1
Failed at 35: sudo -E PATH="$PATH" CRI_RUNTIME="containerd" bash -c "make ksm"
```

**Comment by fidencio** [comment]
KSM failure is hardly related to this PR, and an issue will be opened.

**Comment by fidencio** [comment]
https://github.com/kata-containers/tests/issues/3438

**Comment by Jakob-Naucke** [comment]
> KSM failure is hardly related to this PR, and an issue will be opened.

@fidencio Except it might be.
The KSM test fails similarly on s390x when turning on virtio-fs.
This led me to believe that some change in GoVMM that was merged here introduced this.
However, the KSM test fails on s390x even when applying the bare minimum of https://github.com/kata-containers/govmm/commit/4b136f3f1cc487bd75b2adebff2b06233e318fea#diff-60d10e04edc58668f13d9932884322892005f9b39db600185b0b08872c750579 and otherwise keeping the GoVMM version before this PR.
Note that the changed codepath there does not depend on turning virtio-fs on or off.
This leads me to believe that the breaking change was some time before this PR, and probably in Kata.
However, this is a bit hard to test for me because I would have to test with virtio-fs, but my patches don't always apply cleanly, and I don't currently have an ARM machine at my disposal to run the test myself.
It would be great to know if ARM was already failing KSM before this PR, but unfortunately, the CI results aren't kept so long.

**Comment by Jakob-Naucke** [comment]
Hmmm. When enabling virtio-fs, s390x KSM breaks even when going back as far as October 2020, so maybe the breakage of ARM KSM is unrelated after all.

- Running with `memory-backend-file` on s390x (be it in Kata or in a minimal QEMU) uses no KSM whatsoever for me. 
- Running with `memory-backend-file` in a minimal QEMU on x86_64 gets up to a few pages (120 with two instances, without `memory-backend-file`, it's ~56000, likewise on s390x). In Kata, it's ~2000 (virtio-9p: ~14000).

At one point, the test was also failing on x86_64 for me, but worked again after a reboot.
The test passes on x86_64 even when using `-machine memory-backend=...` instead of `-numa node,memdev=...` (the former being used on s390x because of a lack of DIMM support).

/cc @borntraeger

**Comment by borntraeger** [comment]
@fidenciois  correct. @Jakob-Naucke  KSM only ẃorks for anonymous memory and the testcase tries to see if KSM can find some shareable pages.
So the fact that you use a file backed memory will result in KSM having nothing to find unless there are some other guests or containers. So it might also fail on x86.

**Comment by Jakob-Naucke** [comment]
@fidencio @borntraeger Thank you for your help! Then I take it that KSM tests should simply run with virtio-9p across all architectures. Will fix over at [kata-containers/tests](https://github.com/kata-containers/tests).
